src/Microsoft.AspNetCore.Hosting/ConfigureContainerBuilder.cs

```diff
@@ -0,0 +1,44 @@
+﻿using System;
+using System.Linq;
+using System.Reflection;
+
+namespace Microsoft.AspNetCore.Hosting.Internal
+{
+    public class ConfigureContainerBuilder
+    {
+        public ConfigureContainerBuilder(MethodInfo configureContainerMethod)
+        {
+            MethodInfo = configureContainerMethod;
+        }
+
+        public MethodInfo MethodInfo { get; }
+
+        public Action<object> Build(object instance) => container => Invoke(instance, container);
+
+        private void Invoke(object instance, object container)
+        {
+            if (MethodInfo == null)
+            {
+                return;
+            }
+
+            // Only support IServiceCollection parameters
+            var parameters = MethodInfo.GetParameters();
+            if (parameters.Length != 1 ||
+                parameters.Any(p => !p.ParameterType.IsAssignableFrom(container.GetType())))
+            {
+                throw new InvalidOperationException($"The {MethodInfo.Name} method must either be parameterless or take only one parameter of type {container.GetType()}.");
```

Comment:
It can also be wrong in another way. Consider this scenario:

You have a `IContainerBuilder` interface and derive from `ServiceProviderFactory<IContainerBuilder>`. This class returns an instance of `ContainerBuilder`, which is an internal type.

What you really want the exception message to say is

> The ConfigureContainer method must take only one parameter of type IContainerBuilder.

But because you're using the runtime type, it'll end up saying

> The ConfigureContainer method must take only one parameter of type ContainerBuilder.

Which is slightly misleading because it's an internal type and you'll never be able to take that as a parameter.